### PR TITLE
bpo-44347: [doc] clarify meaning of shutil.copytree's dirs_exist_ok kwarg

### PR DESCRIPTION
**Pull request by jdevries3133**
* It now mirrors the documentation for os.makedirs
* Docstring in python source file was correspondingly updated

https://bugs.python.org/issue44347

https://bugs.python.org/issue44347

### CODE REVIEW

**Comment by jdevries3133** [comment]
@tvogel, continuing our conversation from now-closed https://github.com/python/cpython/pull/26634...

I agree, that should be mentioned and I added more detail for that. Additionally, I moved everything about `dirs_exist_ok` down to the bottom of the documentation and the docstring to match the order of arguments in the function signature.

**Review by nanjekyejoannah — COMMENTED** [review]
Thanks for your contribution @jdevries3133, can you please fix the errors reported in the CI?

**Comment by jdevries3133** [comment]
@nanjekyejoannah I fixed it. Thank you for your patience!

File: Doc/library/shutil.rst
Comment (akulakov) [inline Doc/library/shutil.rst]: I think this phrasing is a bit unclear. 'conversely' implies that the following is opposite to how the default value works. In other words, I would read this sentence as saying that with dirs_exist_ok=False, dirs along the path to dst, as well as dst itself, will not be created.

Comment (jdevries3133) [reply Doc/library/shutil.rst]: Yeah, you are totally correct; I've made it more confusing. I changed it to be much more explicit.

**Comment by jdevries3133** [comment]
I think the failing test is related to [38323](https://bugs.python.org/issue38323). I only made changes to docs / docstrings, and `make check html SPHINXOPTS="-q -W -j4"` was happy with me.

**Comment by terryjreedy** [comment]
@jdevries3133 Do you know what you did a couple of hours ago to balloon the diff by over 100 files and  the review list?  In the past, people have done that by either merging in something other than main or by force-pushing.  I notice that you did a couple main merges 9 days ago that seem correct.  In any case, I made a PR branch, 'git merge upsteam/main', and pushed, and we are back the the correct 3 file changes. (This does not always fix PRs like this but did this time.) (The devguide talks about making an 'upstream' nickname for python/cpython.)

**Comment by jdevries3133** [comment]
@terryjreedy Oh shoot, I'm so sorry. I think what happened was I meant to fast forward my copy of the main branch before checking out onto a new issue, but didn't realize I wasn't on my main branch (I was on *this* branch).

At that point, I knew I had messed up a branch but I didn't notice that I messed up a branch with an open PR against it. My bad, sorry.

File: Doc/library/shutil.rst
Comment (terryjreedy) [inline Doc/library/shutil.rst]: The doc for dirs_exist_ok is misplaced here; moving it to the end us correct.

'or any missing parent directory' is wrong, so it is good that this is changed.  I think the name of the parameteris deceptive and was based on the mistaken understanding.  However, we cannot change it.  The corresponding parameter in os.makedirs is just 'exist_ok', referring to the target directory.

It is normal for parent directories to exist, so that should never raise.  But os.makedirs (plural) also makes intermediate dirs as needed.  This is the difference from os.makedir (singular).  So:
```suggestion
   named *dst* and return the destination directory.   Make all intermediate directories needed to contain *dst*.
```
Unlike os.makedirs, *dst* is not a leaf directory, because it is the root of the copied tree.  Hence the modification of the sentence from os.makedirs.

File: Doc/library/shutil.rst
Comment (terryjreedy) [inline Doc/library/shutil.rst]: The first sentence, copied from the os.makedirs doc, is correct if the latter is.  And this is all that os.makedirs says about the option.

The second sentence appears wrong in that intermediate are always created as needed.  That needs to be said at the top, as suggested in the previous comment.  I don't think more is needed because it should be clear enough that if
```suggestion
   If *dirs_exist_ok* is false (the default) and *dst* already exists, a
   :exc:`FileExistsError` is raised.
```

File: Lib/shutil.py
Comment (terryjreedy) [inline Lib/shutil.py:521]:
```suggestion
    Make all intermediate directories needed to contain *dst*.
    If exception(s) occur, an Error is raised with a list of reasons.
```

File: Lib/shutil.py
Comment (terryjreedy) [inline Lib/shutil.py]:
```suggestion
    If dirs_exist_ok is false (the default) and dst already exists, a
    FileExistsError is raised.
```

**Review by terryjreedy — CHANGES_REQUESTED** [review]

**Comment by terryjreedy** [comment]
@iritkatriel My review is based on your comments on the issue.  I would like it if you could check that my understanding is correct.

**Comment by terryjreedy** [comment]
The macOS failure is in test_ssl, unrelated for sure.

**Comment by iritkatriel** [comment]
@terryjreedy Yes I agree with your correction (I tested it and it creates the missing parent directories regardless of the value of dirs_exist_ok).

@jdevries3133 Off topic, but while we're here I spotted a typo that you could fix as well: 

`Changed in version 3.2: Added the copy_function argument to be able to provide a custom copy function. Added the ignore_dangling_symlinks argument to silent dangling symlinks errors when symlinks is false.`

Should be "silence" rather than "silent".

**Comment by jdevries3133** [comment]
@iritkatriel 82ae9d6f11282ccbb8897062cac6ab93e72acd97 sneaked in here somehow. I do not know where it came from or how, but naïvely, this is how I would fix it:

```bash
git reset --hard a0a46be8ae76f4e671e69c7a984af5bb9c572d5e  # last of the suggested changes from terry
git push --force
```

I'm not sure what that will do to this already-ailing PR branch, though... what do you think? Thank you for your support, I'm sorry this has gotten a bit out of whack.

**Comment by terryjreedy** [comment]
Leave the spelling fix.  The bogus addition is e02e998, with was the only merge to main since earlier.  I don't know how it appeared, but the same `git merge upstream/main` I did before fixed it.  Force pushes is how people mess up PRs.

File: Misc/NEWS.d/next/Documentation/2020-06-10-16-55-55.bpo-44347.Fjvuw7.rst
Comment (akulakov) [inline Misc/NEWS.d/next/Documentation/2020-06-10-16-55-55.bpo-44347.Fjvuw7.rst]: Let's make it a link -- :func:`shutil.copytree`

**Comment by akulakov** [comment]
I agree with @terryjreedy that the arg name is misleading, perhaps a note should be added to this effect? Otherwise it may be extra confusing because `copytree` does create multiple nested directories, and the user might think the arg name refers to that.

**Comment by jdevries3133** [comment]
@akulakov I added a note, what do you think?

Also, "I have made the requested changes; please review again."

**Comment by akulakov** [comment]
> @akulakov I added a note, what do you think?

Sorry I should have added a suggestion to my comment, I was thinking of something shorter, along the lines:

`Note that the name of the arg *dirs_exist_ok* may be misleading, the effect of the argument is similar to :func:`os.makedirs` 's *exist_ok*, which might have been a more accurate name.`

My reasoning is that if there is a mismatch between the arg name and the docs, a user will wonder which one to trust more and will probably end up doing their own testing, so in this case it may be good to just confess to it (the mismatch) in the docs :-). 

However let's keep your version for now and wait for @terryjreedy to do a final review pass.

**Comment by jdevries3133** [comment]
Ah, I see what you are saying; however, I'm still happy with what I added because:

1. *dirs_exist_ok* is a good name for the parameter here; simply *exist_ok* would not be a better name. In `os.makedirs(exist_ok=False)`, it is clear to the user that this switch only applies to directories (because makedirs itself obviously only touches directories). `shutil.copytree`'s function name itself does not provide the same context, though, so it is worthwhile for the parameter to be more explicit.
2. Referencing `os.makedirs` will only help users who already know how that function behaves. For newcomers, I don't want to give only a breadcrumb and send them on a scavenger hunt.

All that being said, I did add one sentence about the relationship between *dirs_exist_ok* and os.makedirs's *exist_ok* parameters.

File: Doc/library/shutil.rst
Comment (JelleZijlstra) [inline Doc/library/shutil.rst]: Could this mention how the behavior depends on whether `dir_exist_ok` is passed? As written the example doesn't explicitly explain where it makes a difference.

Comment (JelleZijlstra) [reply Doc/library/shutil.rst]: Also, this example feels a bit excessive; we generally aim for concision in the docs. I see that lots of people have already contributed to this discussion though, so if they're in general agreement I'm OK with adding it.

Comment (jdevries3133) [reply Doc/library/shutil.rst]: Yeah, I'm not crazy about the lack of brevity in it either, anymore. This was one of my first contributions and I was feeling myself a bit too much –– 

I will fix this up sometime soon when I get a chance. Thank you for the ping.

**Comment by jdevries3133** [comment]
@JelleZijlstra I revised the PR. If it looks good, I will add the paragraph about *dirs_exist_ok* to the docstring in `Lib/shutil.py`, and then this should be good to go.

I deleted 80% of the PR and rewrote the rest, so please review as if this is all new.

File: Doc/library/shutil.rst
Comment (JelleZijlstra) [inline Doc/library/shutil.rst]: "kwarg" is a bit too informal. Maybe the parenthetical should just be removed?

File: Doc/library/shutil.rst
Comment (JelleZijlstra) [inline Doc/library/shutil.rst]: We should avoid making promises about the implementation. Maybe just remove this sentence.

**Review by JelleZijlstra — COMMENTED** [review]
Thanks!

File: Doc/library/shutil.rst
Comment (jdevries3133) [reply Doc/library/shutil.rst]: I'm ok with that, it's removed in c169b1d3d2d67e9be37a95b3843896b0bd3070d9

File: Doc/library/shutil.rst
Comment (jdevries3133) [reply Doc/library/shutil.rst]: Ok, I removed that sentence in c169b1d3d2d67e9be37a95b3843896b0bd3070d9

**Review by JelleZijlstra — COMMENTED** [review]
The text looks good! Can you update the docstring too and undraft the PR?

**Comment by jdevries3133** [comment]
@JelleZijlstra OK, good to go. Not sure why the CLA bot is complaining, I clicked the link and re-signed but it still says I'm unsigned. I'm guessing it's just a GitHub migration glitch, I've made [a handful of contributions in the past](https://github.com/python/cpython/commits/main?author=jdevries3133) and did sign the CLA.

**Comment by jdevries3133** [comment]
CI failure is a fluke:

```
E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/c/ccache/ccache_3.7.7-1_amd64.deb  Connection failed [IP: 52.147.219.192 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
Error: Process completed with exit code 100.
```

Can you trigger a re-run?

**Comment by AlexWaygood** [comment]
> CI failure is a fluke:
> 
> ```
> E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/c/ccache/ccache_3.7.7-1_amd64.deb  Connection failed [IP: 52.147.219.192 80]
> E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
> Error: Process completed with exit code 100.
> ```
> 
> Can you trigger a re-run?

Probably related to https://github.com/python/the-knights-who-say-ni/pull/318

**Review by JelleZijlstra — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
I wonder if the CLA bot is unhappy because of all the random commits that were added to the branch earlier.

**Comment by AlexWaygood** [comment]
> I wonder if the CLA bot is unhappy because of all the random commits that were added to the branch earlier

Perhaps a force-push would sort that out?

**Comment by jdevries3133** [comment]
I just created a new branch and a new PR with a single commit:

https://github.com/python/cpython/pull/91434

Use it if that's helpful, close it if not :)

**Comment by JelleZijlstra** [comment]
Merged in #91434.

**Pull request closed without merging**
